### PR TITLE
Support the `jitter` function in the constant policy

### PR DESCRIPTION
**Pull request by moznion**
毎度度々すみません。

Currently, the constant policy doesn't support the jitter mechanism, but sometimes I'd like to use such a feature. For example, it doesn't need to extend the duration according to the number of retries but in some cases, it requires jitter to avoid sort of floodings.

This pull-request applies the following changes:

- `WithJitterFactor` and `WithRNG` returns `JitterOption`.
  - to use those options by the exponential policy and constant one.
- Introducing jitter mechanism into the constant policy.
- Make `jitter` be strategies.
  - `jitter#apply()` is always applied to the interval value in a policy, but it chooses which jitter strategy to use.
    - `randomJitter#apply()` does the same as the current jitter implementation
    - `notJitter#apply()` does nothing

### CODE REVIEW

File: v2/options.go [hunk 1]
```diff
@@ -55,6 +57,22 @@ type exponentialOption struct {
 
 func (*exponentialOption) exponentialOption() {}
 
+// JitterOption is an option that is used by policy that supports jitter.
+type JitterOption interface {
```
Comment (lestrrat) [inline v2/options.go:61]: Let's make this `CommonOption`

```
// CommonOption is an option that can be passed to any of the  backoff policies.
```

File: v2/options.go [hunk 2]
```diff
@@ -55,6 +57,22 @@ type exponentialOption struct {
...
+	Option
```
Comment (lestrrat) [inline v2/options.go:62]: This is redundant. It's safe to delete it

File: v2/options.go [hunk 3]
```diff
@@ -55,6 +57,22 @@ type exponentialOption struct {
...
+	ExponentialOption
+	ConstantOption
+	jitterOption()
```
Comment (lestrrat) [inline v2/options.go:65]: If we make it a `CommonOption`, we don't need this

File: v2/options.go [hunk 4]
```diff
@@ -55,6 +57,22 @@ type exponentialOption struct {
...
+}
+
+type jitterOption struct {
```
Comment (lestrrat) [inline v2/options.go:68]: We won't need this either

File: v2/options.go [hunk 5]
```diff
@@ -100,14 +118,14 @@ func WithMultiplier(v float64) ExponentialOption {
 // value outside of this range is specified, the value will be silently
 // ignored and jittering is disabled.
 //
-// This option can be passed to ExponentialPolicy constructor
-func WithJitterFactor(v float64) ExponentialOption {
-	return &exponentialOption{option.New(identJitterFactor{}, v)}
+// This option can be passed to ExponentialPolicy or ConstantPolicy constructor
+func WithJitterFactor(v float64) JitterOption {
```
Comment (lestrrat) [inline v2/options.go:122]: return `CommonOption`

File: v2/options.go [hunk 6]
```diff
@@ -100,14 +118,14 @@ func WithMultiplier(v float64) ExponentialOption {
...
+	return &jitterOption{option.New(identJitterFactor{}, v)}
```
Comment (lestrrat) [inline v2/options.go:123]: `return &exponentialOption{constantOption{option.New(identJitterFactor)}}` should suffice

File: v2/options.go [hunk 7]
```diff
@@ -100,14 +118,14 @@ func WithMultiplier(v float64) ExponentialOption {
...
 }
 
 // WithRNG specifies the random number generator used for jittering.
 // If not provided one will be created, but if you want a truly random
 // jittering, make sure to provide one that you explicitly initialized
-func WithRNG(v Random) ExponentialOption {
-	return &exponentialOption{option.New(identRNG{}, v)}
+func WithRNG(v Random) JitterOption {
+	return &jitterOption{option.New(identRNG{}, v)}
```
Comment (lestrrat) [inline v2/options.go:130]: Same here.

File: v2/nop_jitter.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package backoff
```
Comment (lestrrat) [inline v2/nop_jitter.go:1]: Please put this in the same `jitter.go` fiile

**Review by lestrrat — CHANGES_REQUESTED** [review]
I don't like the fact that it's not immediately obvious who can accept a JitterOption.
let's just call it `CommonOption`. We'll worry about policies that don't support jittering later. (NullPolicy is an exception. It's obvious it doesn't do anything)

File: v2/options.go (hunk 6 above)
Comment (moznion) [reply v2/options.go:123]:
```go
type CommonOption interface {
	ExponentialOption
	ConstantOption
}

func WithJitterFactor(v float64) CommonOption {
	return &exponentialOption{constantOption{option.New(identJitterFactor)}}
}
```

In this case, go compiler complains like the following, hmm:

```
Cannot use '&exponentialOption{constantOption{option.New(identJitterFactor)}}' (type *exponentialOption) as type CommonOption Type does not implement 'CommonOption' as some methods are missing: constantOption()
```

Comment (lestrrat) [reply v2/options.go:123]: Ah, oops. then I guess it's back to

```go
type CommonOption interface {
    ExponentialOption
    ConstantOption
}

type commonOption struct { Option }
func (*commonOption) exponentialOption() {}
func (*commonOption) constantOption() {}

func WithJitterFactor(v float64) CommonOption { 
  return &commonOption{option.New(...)}
}
```

File: v2/nop_jitter.go (hunk 8 above)
Comment (moznion) [reply v2/nop_jitter.go:1]: Do you prefer to move also `random_jitter`?

Comment (lestrrat) [reply v2/nop_jitter.go:1]: I probably just didn't notice. yeah, just put it in jitter.go, please.

File: v2/options.go (hunk 6 above)
Comment (moznion) [reply v2/options.go]: the snapshot of the current status is here: https://github.com/lestrrat-go/backoff/pull/25/commits/b5cd454a34a569b47545f4b74ae35280575399d1

Comment (moznion) [reply v2/options.go]: yeah, we are on the same page.

**Review by lestrrat — APPROVED** [review]

**Comment by moznion** [comment]
@lestrrat Thank you for your review! I've changed the code according to your suggestions.

**Comment by lestrrat** [comment]
merged! thanks

**Comment by moznion** [comment]
Thank you, too!

**Pull request merged**